// @stacks/transactions/dist/clarity/values/principalCV.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.principalCV = principalCV;
exports.standardPrincipalCV = standardPrincipalCV;
exports.standardPrincipalCVFromAddress = standardPrincipalCVFromAddress;
exports.contractPrincipalCV = contractPrincipalCV;
exports.contractPrincipalCVFromAddress = contractPrincipalCVFromAddress;
exports.contractPrincipalCVFromStandard = contractPrincipalCVFromStandard;
const common_1 = require("@stacks/common");
const wire_1 = require("../../wire");
const constants_1 = require("../constants");
function principalCV(principal) {
    if (principal.includes('.')) {
        const [address, contractName] = principal.split('.');
        return contractPrincipalCV(address, contractName);
    }
    else {
        return standardPrincipalCV(principal);
    }
}
function standardPrincipalCV(addressString) {
    const addr = (0, wire_1.createAddress)(addressString);
    return { type: constants_1.ClarityType.PrincipalStandard, value: (0, wire_1.addressToString)(addr) };
}
function standardPrincipalCVFromAddress(address) {
    return { type: constants_1.ClarityType.PrincipalStandard, value: (0, wire_1.addressToString)(address) };
}
function contractPrincipalCV(addressString, contractName) {
    const addr = (0, wire_1.createAddress)(addressString);
    const lengthPrefixedContractName = (0, wire_1.createLPString)(contractName);
    return contractPrincipalCVFromAddress(addr, lengthPrefixedContractName);
}
function contractPrincipalCVFromAddress(address, contractName) {
    if ((0, common_1.utf8ToBytes)(contractName.content).byteLength >= 128) {
        throw new Error('Contract name must be less than 128 bytes');
    }
    return {
        type: constants_1.ClarityType.PrincipalContract,
        value: `${(0, wire_1.addressToString)(address)}.${contractName.content}`,
    };
}
function contractPrincipalCVFromStandard(sp, contractName) {
    return {
        type: constants_1.ClarityType.PrincipalContract,
        value: `${sp.value}.${contractName}`,
    };
}
