// @stacks/transactions/dist/postcondition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePostConditionAmount = parsePostConditionAmount;
exports.postConditionToWire = postConditionToWire;
exports.wireToPostCondition = wireToPostCondition;
exports.conditionTypeToByte = conditionTypeToByte;
exports.conditionByteToType = conditionByteToType;
exports.poxConditionTypeToByte = poxConditionTypeToByte;
exports.poxConditionByteToType = poxConditionByteToType;
exports.postConditionToHex = postConditionToHex;
exports.postConditionModeFrom = postConditionModeFrom;
const common_1 = require("@stacks/common");
const constants_1 = require("./constants");
const wire_1 = require("./wire");
const MAX_U64 = BigInt('18446744073709551615');
function parsePostConditionAmount(amount) {
    const value = (0, common_1.intToBigInt)(amount);
    if (value < 0n || value > MAX_U64) {
        throw new RangeError(`Post-condition amount must be between 0 and ${MAX_U64} (u64 max), received: ${value}`);
    }
    return value;
}
var PostConditionCodeWireType;
(function (PostConditionCodeWireType) {
    PostConditionCodeWireType[PostConditionCodeWireType["eq"] = 1] = "eq";
    PostConditionCodeWireType[PostConditionCodeWireType["gt"] = 2] = "gt";
    PostConditionCodeWireType[PostConditionCodeWireType["lt"] = 4] = "lt";
    PostConditionCodeWireType[PostConditionCodeWireType["gte"] = 3] = "gte";
    PostConditionCodeWireType[PostConditionCodeWireType["lte"] = 5] = "lte";
    PostConditionCodeWireType[PostConditionCodeWireType["sent"] = 16] = "sent";
    PostConditionCodeWireType[PostConditionCodeWireType["not-sent"] = 17] = "not-sent";
    PostConditionCodeWireType[PostConditionCodeWireType["maybe-sent"] = 18] = "maybe-sent";
})(PostConditionCodeWireType || (PostConditionCodeWireType = {}));
var PoxConditionCodeWireType;
(function (PoxConditionCodeWireType) {
    PoxConditionCodeWireType[PoxConditionCodeWireType["will-not-perform"] = 48] = "will-not-perform";
    PoxConditionCodeWireType[PoxConditionCodeWireType["may-perform"] = 49] = "may-perform";
    PoxConditionCodeWireType[PoxConditionCodeWireType["will-perform"] = 50] = "will-perform";
})(PoxConditionCodeWireType || (PoxConditionCodeWireType = {}));
function postConditionToWire(postcondition) {
    switch (postcondition.type) {
        case 'stx-postcondition':
            return {
                type: wire_1.StacksWireType.PostCondition,
                conditionType: constants_1.PostConditionType.STX,
                principal: postcondition.address === 'origin'
                    ? { type: wire_1.StacksWireType.Principal, prefix: constants_1.PostConditionPrincipalId.Origin }
                    : (0, wire_1.parsePrincipalString)(postcondition.address),
                conditionCode: conditionTypeToByte(postcondition.condition),
                amount: parsePostConditionAmount(postcondition.amount),
            };
        case 'ft-postcondition':
            return {
                type: wire_1.StacksWireType.PostCondition,
                conditionType: constants_1.PostConditionType.Fungible,
                principal: postcondition.address === 'origin'
                    ? { type: wire_1.StacksWireType.Principal, prefix: constants_1.PostConditionPrincipalId.Origin }
                    : (0, wire_1.parsePrincipalString)(postcondition.address),
                conditionCode: conditionTypeToByte(postcondition.condition),
                amount: parsePostConditionAmount(postcondition.amount),
                asset: (0, wire_1.parseAssetString)(postcondition.asset),
            };
        case 'nft-postcondition':
            return {
                type: wire_1.StacksWireType.PostCondition,
                conditionType: constants_1.PostConditionType.NonFungible,
                principal: postcondition.address === 'origin'
                    ? { type: wire_1.StacksWireType.Principal, prefix: constants_1.PostConditionPrincipalId.Origin }
                    : (0, wire_1.parsePrincipalString)(postcondition.address),
                conditionCode: conditionTypeToByte(postcondition.condition),
                asset: (0, wire_1.parseAssetString)(postcondition.asset),
                assetName: postcondition.assetId,
            };
        case 'staking-postcondition':
            return {
                type: wire_1.StacksWireType.PostCondition,
                conditionType: constants_1.PostConditionType.Staking,
                principal: postcondition.address === 'origin'
                    ? { type: wire_1.StacksWireType.Principal, prefix: constants_1.PostConditionPrincipalId.Origin }
                    : (0, wire_1.parsePrincipalString)(postcondition.address),
                conditionCode: conditionTypeToByte(postcondition.condition),
                amount: parsePostConditionAmount(postcondition.amount),
            };
        case 'pox-postcondition':
            return {
                type: wire_1.StacksWireType.PostCondition,
                conditionType: constants_1.PostConditionType.PoX,
                principal: postcondition.address === 'origin'
                    ? { type: wire_1.StacksWireType.Principal, prefix: constants_1.PostConditionPrincipalId.Origin }
                    : (0, wire_1.parsePrincipalString)(postcondition.address),
                conditionCode: poxConditionTypeToByte(postcondition.condition),
            };
        default:
            throw new Error('Invalid post condition type');
    }
}
function wireToPostCondition(wire) {
    switch (wire.conditionType) {
        case constants_1.PostConditionType.STX:
            return {
                type: 'stx-postcondition',
                address: principalWireToString(wire.principal),
                condition: conditionByteToType(wire.conditionCode),
                amount: wire.amount.toString(),
            };
        case constants_1.PostConditionType.Fungible:
            return {
                type: 'ft-postcondition',
                address: principalWireToString(wire.principal),
                condition: conditionByteToType(wire.conditionCode),
                amount: wire.amount.toString(),
                asset: assetWireToString(wire.asset),
            };
        case constants_1.PostConditionType.NonFungible:
            return {
                type: 'nft-postcondition',
                address: principalWireToString(wire.principal),
                condition: conditionByteToType(wire.conditionCode),
                asset: assetWireToString(wire.asset),
                assetId: wire.assetName,
            };
        case constants_1.PostConditionType.Staking:
            return {
                type: 'staking-postcondition',
                address: principalWireToString(wire.principal),
                condition: conditionByteToType(wire.conditionCode),
                amount: wire.amount.toString(),
            };
        case constants_1.PostConditionType.PoX:
            return {
                type: 'pox-postcondition',
                address: principalWireToString(wire.principal),
                condition: poxConditionByteToType(wire.conditionCode),
            };
        default: {
            const _exhaustiveCheck = wire;
            throw new Error(`Invalid post condition type: ${_exhaustiveCheck}`);
        }
    }
}
function conditionTypeToByte(condition) {
    return PostConditionCodeWireType[condition];
}
function conditionByteToType(wireType) {
    return PostConditionCodeWireType[wireType];
}
function poxConditionTypeToByte(condition) {
    return PoxConditionCodeWireType[condition];
}
function poxConditionByteToType(wireType) {
    return PoxConditionCodeWireType[wireType];
}
function postConditionToHex(postcondition) {
    const wire = postConditionToWire(postcondition);
    return (0, wire_1.serializePostConditionWire)(wire);
}
function postConditionModeFrom(mode) {
    if (typeof mode === 'number')
        return mode;
    if (mode === 'allow')
        return constants_1.PostConditionMode.Allow;
    if (mode === 'deny')
        return constants_1.PostConditionMode.Deny;
    if (mode === 'originator')
        return constants_1.PostConditionMode.Originator;
    throw new Error(`Invalid post condition mode: ${mode}`);
}
function assetWireToString(asset) {
    const address = (0, wire_1.addressToString)(asset.address);
    const contractId = `${address}.${asset.contractName.content}`;
    return `${contractId}::${asset.assetName.content}`;
}
function principalWireToString(principal) {
    switch (principal.prefix) {
        case constants_1.PostConditionPrincipalId.Origin:
            return 'origin';
        case constants_1.PostConditionPrincipalId.Standard:
            return (0, wire_1.addressToString)(principal.address);
        case constants_1.PostConditionPrincipalId.Contract:
            const address = (0, wire_1.addressToString)(principal.address);
            return `${address}.${principal.contractName.content}`;
        default:
            const _exhaustiveCheck = principal;
            throw new Error(`Invalid principal type: ${_exhaustiveCheck}`);
    }
}
